cmd/promtool/backfill.go

```diff
@@ -106,15 +115,24 @@ func createBlocks(input []byte, mint, maxt int64, maxSamplesInAppender int, outp
 					continue
 				}
 
-				l := labels.Labels{}
-				p.Metric(&l)
 				_, ts, v := p.Series()
 				if ts == nil {
+					l := labels.Labels{}
+					p.Metric(&l)
 					return errors.Errorf("expected timestamp for series %v, got none", l)
 				}
-				if *ts < t || *ts >= tsUpper {
+				if *ts < t {
 					continue
 				}
+				if *ts >= tsUpper {
+					if *ts < nextSampleTs {
+						nextSampleTs = *ts
+					}
+					continue
```

Comment:
I think this has to be `break`, and you should set `nextSampleTs` to `MinInt64` initially, and then you don't need a few lines of code above. It's probably easiest to show you in the code. I cannot just add a commit to your branch because it lives in your own GH org. I'll create a PR on top of this one.

Reply:
You can because you are a maintainer of this repository

Reply:
https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Reply:
TIL...

Reply:
And see discussion in other PR, you are actually right with your flow here.